src/util/json_stream.cpp

```diff
@@ -0,0 +1,153 @@
+
+#include "json_stream.h"
+
+#include <ostream>
+
+/// Output the element delimiter to the output stream
+void json_streamt::output_delimiter()
+{
+  if(!first)
+    out << ',';
+  else
+    first = false;
+  out << '\n';
+  out << std::string((indent + 1) * 2, ' ');
+}
+
+/// Construct a new JSON array stream
+/// \param out: output stream
+/// \param indent: indentation level
+json_stream_arrayt::json_stream_arrayt(std::ostream &out, unsigned indent)
+  : json_streamt(out, indent)
+{
+  out << '[';
+}
+
+/// Output the non-streaming JSON objects and closes the current child stream
+void json_stream_arrayt::output_child_stream()
+{
+  if(!object.empty())
+  {
+    output_delimiter();
+    object["array_element"].output_rec(out, indent + 1);
+    object.clear();
+  }
+  if(child_stream)
+  {
+    child_stream->close();
+  }
+}
+
+/// Output the finalizing character for a JSON array
+void json_stream_arrayt::output_finalizer()
+{
+  out << '\n' << std::string(indent * 2, ' ');
+  out << ']';
+}
+
+/// Constructor for json_stream_objectt.
+/// \param out: The stream that is to be used to output the JSON object.
+/// \param indent: Current indentation level.
+json_stream_objectt::json_stream_objectt(std::ostream &out, unsigned indent)
+  : json_streamt(out, indent)
+{
+  out << '{';
+}
+
+/// Create a new JSON array child stream.
+json_stream_arrayt &json_streamt::create_child_stream_array()
+{
+  child_stream =
+    std::unique_ptr<json_streamt>(new json_stream_arrayt(out, indent + 1));
+  return static_cast<json_stream_arrayt &>(*child_stream);
+}
+
+/// Create a new JSON object child stream.
+json_stream_objectt &json_streamt::create_child_stream_object()
+{
+  child_stream =
+    std::unique_ptr<json_streamt>(new json_stream_objectt(out, indent + 1));
+  return static_cast<json_stream_objectt &>(*child_stream);
+}
+
+/// Add a JSON object child stream
+json_stream_objectt &json_stream_arrayt::push_back_stream_object()
+{
+  PRECONDITION(open);
+  // To ensure consistency of the output, we flush and
+  // close the current child stream before creating the new one.
+  output_child_stream();
+  output_delimiter();
+  return create_child_stream_object();
+}
+
+/// Add a JSON array child stream
+json_stream_arrayt &json_stream_arrayt::push_back_stream_array()
+{
+  PRECONDITION(open);
+  // To ensure consistency of the output, we flush and
+  // close the current child stream before creating the new one.
+  output_child_stream();
+  output_delimiter();
+  return create_child_stream_array();
+}
+
+/// Add a JSON object stream for a specific key
+/// \param key: key of the JSON property
+json_stream_objectt &
+json_stream_objectt::push_back_stream_object(const std::string &key)
+{
+  PRECONDITION(open);
+  // To ensure consistency of the output, we flush and
+  // close the current child stream before creating the new one.
+  output_child_stream();
+  output_delimiter();
+  jsont::output_key(out, key);
+  return create_child_stream_object();
+}
```

Comment:
What is the purpose of all these `push_back_stream*` functions ("Add a JSON ..." -- to what?) as none of them are invoked anywhere in this commit, and why aren't they just a single one in the parent class? I understand that they have different return types, but is that actually necessary (as anything done to those return values can be resolved to the right child class via virtual tables)?

Reply:
I think this fits with my overall comment: there seems to be a strong case for re-architecting here to use a visitor rather than this proposed design of inheritance.

Reply:
@tautschnig, could you please explain how 'use a visitor' would contribute to a solution in this context?
Inheritance is used here because json_stream_arrayt and json_stream_objectt are both json streams, which share common functionality, namely flushing the data. If this is considered too little to justify a base class the common code can be duplicated in the respective derived classes at the cost of complicating the logic to flush the stream and to enforce the single-stream-at-a-time invariant.

Reply:
In my opinion, there should be one class that knows how to stream `jsont`. That class would be the visitor. It has methods that know how to stream a `json_arrayt` or `json_objectt` (and `json_numbert`, `json_stringt`, etc.).